astropy/utils/masked/core.py

```diff
@@ -1055,38 +1055,42 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=False):
                 dtype = np.dtype('f4')
                 is_float16_result = out is None
 
+        where = ~self.mask & where
+
         result = self.sum(axis=axis, dtype=dtype, out=out,
-                          keepdims=keepdims, where=~self.mask)
-        n = np.add.reduce(~self.mask, axis=axis, keepdims=keepdims)
+                          keepdims=keepdims, where=where)
+        n = np.add.reduce(where, axis=axis, keepdims=keepdims)
         result /= n
         if is_float16_result:
             result = result.astype(self.dtype)
         return result
 
-    def var(self, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    def var(self, axis=None, dtype=None, out=None, ddof=0, keepdims=False, *, where=True):
+        where_final = ~self.mask & where
+
         # Simplified implementation based on that in numpy/core/_methods.py
-        n = np.add.reduce(~self.mask, axis=axis, keepdims=keepdims)[...]
+        n = np.add.reduce(where_final, axis=axis, keepdims=keepdims)[...]
 
         # Cast bool, unsigned int, and int to float64 by default.
         if dtype is None and issubclass(self.dtype.type,
                                         (np.integer, np.bool_)):
             dtype = np.dtype('f8')
-        mean = self.mean(axis=axis, dtype=dtype, keepdims=True)
+        mean = self.mean(axis=axis, dtype=dtype, keepdims=True, where=where)
 
         x = self - mean
         x *= x.conjugate()  # Conjugate just returns x if not complex.
 
         result = x.sum(axis=axis, dtype=dtype, out=out,
-                       keepdims=keepdims, where=~x.mask)
```

Comment:
@mhvk, our conversation above makes me wonder why the mask had to be passed into `sum()` explicitly here.
Isn't `x` a `MaskedNDArray`, making the use of the `where` keyword argument redundant in the original implementation?

If we decide it is redundant, I need to change the patch to be `where=where` instead of `where=where_final`

Reply:
By default, the mask is propagated in `.sum()` so here it has to be passed explicitly that we want to ignore masked elements. This is one of the differences with `np.ma.MaskedArray`; for the rationale, see https://docs.astropy.org/en/latest/utils/masked/index.html#utils-masked-vs-numpy-maskedarray

p.s. Hope to be able to review soon, but things are a bit hectic right now.

Reply:
Thanks for showing me that documentation, I was not aware of that difference.

No worries on the review, and thanks for the feedback that you've already provided.